common/coconut-interface/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+    ) -> VerifyCredentialBody {
+        VerifyCredentialBody {
+            n_params,
+            theta: theta.clone(),
+            public_attributes: public_attributes
+                .iter()
+                .map(|attr| attr.to_bs58())
+                .collect(),
+        }
+    }
+
+    pub fn public_attributes(&self) -> Vec<Attribute> {
+        self.public_attributes
+            .iter()
+            .map(|x| Attribute::try_from_bs58(x).unwrap())
+            .collect()
+    }
+}
+//  All strings are base58 encoded representations of structs
+#[derive(Serialize, Deserialize, Debug, Getters, CopyGetters)]
+pub struct BlindSignRequestBody {
+    #[getset(get = "pub")]
+    blind_sign_request: BlindSignRequest,
+    #[getset(get = "pub")]
+    public_key: PublicKey,
+    public_attributes: Vec<String>,
+    #[getset(get = "pub")]
+    total_params: u32,
+}
+
+impl BlindSignRequestBody {
+    pub fn new(
+        blind_sign_request: &BlindSignRequest,
+        public_key: &PublicKey,
+        public_attributes: &[Attribute],
+        total_params: u32,
+    ) -> BlindSignRequestBody {
+        BlindSignRequestBody {
+            blind_sign_request: blind_sign_request.clone(),
+            public_key: public_key.clone(),
+            public_attributes: public_attributes
+                .iter()
+                .map(|attr| attr.to_bs58())
+                .collect(),
+            total_params,
+        }
+    }
+
+    pub fn public_attributes(&self) -> Vec<Attribute> {
+        self.public_attributes
+            .iter()
+            .map(|x| Attribute::try_from_bs58(x).unwrap())
+            .collect()
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct BlindedSignatureResponse {
+    pub blinded_signature: BlindedSignature,
+}
+
+impl BlindedSignatureResponse {
+    pub fn new(blinded_signature: BlindedSignature) -> BlindedSignatureResponse {
+        BlindedSignatureResponse { blinded_signature }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct VerificationKeyResponse {
+    pub key: VerificationKey,
+}
+
+impl VerificationKeyResponse {
+    pub fn new(key: VerificationKey) -> VerificationKeyResponse {
+        VerificationKeyResponse { key }
+    }
+}
+
+fn get_verification_key(url: &str) -> Result<VerificationKey, String> {
+    match attohttpc::get(format!("{}/v1/verification_key", url)).send() {
+        Ok(resp) => {
+            let verification_key_response: VerificationKeyResponse = resp.json().unwrap();
+            Ok(verification_key_response.key)
+        }
+        Err(e) => Err(format!("{}", e)),
+    }
+}
+
+pub fn get_aggregated_verification_key(
+    validator_urls: Vec<String>,
+) -> Result<VerificationKey, String> {
+    let mut verification_keys = Vec::new();
+    let mut indices = Vec::new();
+
+    for (idx, url) in validator_urls.iter().enumerate() {
+        verification_keys.push(get_verification_key(url.as_ref())?);
+        indices.push((idx + 1) as u64);
```

Comment:
The problem with the sucker is that you must ensure your validator urls are in correct orders otherwise you will mess up your indices (I guess it should rather be called `points` as technically those are x-coordinates of a polynomial used during key generation). Perhaps we should slightly modify `coconut` library itself so that the coordinate/index would already be included in the serialized verification key?

Reply:
I agree we should unite the two, but that should be done separately, probably in the `coconut-rs` library.